substrate/frame/support/src/tests/custom_default_config.rs

```diff
@@ -0,0 +1,64 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::Config;
+use crate::{derive_impl, pallet_prelude::inject_runtime_type};
+use static_assertions::assert_type_eq_all;
+
+pub mod custom_config_prelude {
+	use super::*;
+
+	#[docify::export(custom_default_trait)]
+	pub trait CustomDefaultConfigProvider {
+		type AccountId;
+		type RuntimeOrigin;
+		type RuntimeCall;
+		type RuntimeTask;
+	}
+
+	pub struct CustomDefaultConfig;
+
+	#[docify::export(custom_default_impl)]
+	#[crate::register_default_impl(CustomDefaultConfig)]
+	impl CustomDefaultConfigProvider for CustomDefaultConfig {
+		type AccountId = u16;
+		#[inject_runtime_type]
+		type RuntimeOrigin = ();
+		#[inject_runtime_type]
+		type RuntimeCall = ();
+		#[inject_runtime_type]
+		type RuntimeTask = ();
+	}
+}
+
+#[docify::export(custom_default_derive)]
+#[test]
+fn derive_impl_works_with_custom_default_config() {
+	struct DummyRuntime;
+
+	#[derive_impl(custom_config_prelude::CustomDefaultConfig, no_aggregated_types)]
+	impl Config for DummyRuntime {
+		type BaseCallFilter = frame_support::traits::Everything;
+		type Block = super::Block;
+		type DbWeight = ();
+		type PalletInfo = super::PalletInfo;
+	}
+
+	assert_type_eq_all!(<DummyRuntime as Config>::AccountId, u16);
+	assert_type_eq_all!(<DummyRuntime as Config>::RuntimeOrigin, ());
+	assert_type_eq_all!(<DummyRuntime as Config>::RuntimeCall, ());
+}
```

Comment:
How about:

1. One example that demonstrate how this can be used in a pallet, for example to solve the issue of @seadanda. 
2. (This is the one I am more interested in) One example of how this can be used in a standalone trait. For example, to build a custom trait like: 

```
trait OpinionatedSystem: frame_system::Config { // more types and fns.}
// register default for this. 

#[frame::pallet] mod pallet {
    #[pallet::config]
    pub trait Config: OpinionatedSystem { .. }
}

mod runtime {
    // Use the default provided
    impl pallet::Config for Runtime { .. }
}
```

And please focus on making both as self-explanatory and e2e as possible?